internal/fileutil/fileutil.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package fileutil
+
+import (
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+)
+
+// CreateAndSyncFileAtomically writes the content to the tmpFile, fsyncs the tmpFile,
+// renames the tmpFile to the finalFile. In other words, in the event of a crash, either
+// the final file will not be visible or it will have the full contents.
+// The tmpFile should not be existing. The finalFile, if exists, will be overwritten (default rename behavior)
+func CreateAndSyncFileAtomically(dir, tmpFile, finalFile string, content []byte, perm os.FileMode) error {
+	tempFilePath := filepath.Join(dir, tmpFile)
+	finalFilePath := filepath.Join(dir, finalFile)
+	if err := CreateAndSyncFile(tempFilePath, content, perm); err != nil {
+		return err
+	}
+	if err := os.Rename(tempFilePath, finalFilePath); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CreateAndSyncFile creates a file, writes the content and syncs the file
+func CreateAndSyncFile(filePath string, content []byte, perm os.FileMode) error {
+	file, err := os.OpenFile(filePath, os.O_RDWR|os.O_CREATE|os.O_EXCL, perm)
+	if err != nil {
+		return errors.Wrapf(err, "error while creating file:%s", filePath)
+	}
+	_, err = file.Write(content)
+	if err != nil {
+		file.Close()
+		return errors.Wrapf(err, "error while writing to file:%s", filePath)
+	}
+	if err = file.Sync(); err != nil {
+		file.Close()
+		return errors.Wrapf(err, "error while synching the file:%s", filePath)
+	}
+	if err := file.Close(); err != nil {
+		return errors.Wrapf(err, "error while closing the file:%s", filePath)
+	}
+	return nil
+}
+
+// SyncDir fsyncs the given dir
+func SyncDir(dirPath string) error {
+	dir, err := os.Open(dirPath)
+	if err != nil {
+		return errors.Wrapf(err, "error while opening dir:%s", dirPath)
+	}
+	if err := dir.Sync(); err != nil {
+		dir.Close()
+		return errors.Wrapf(err, "error while synching dir:%s", dirPath)
+	}
+	if err := dir.Close(); err != nil {
+		return errors.Wrapf(err, "error while closing dir:%s", dirPath)
+	}
+	return err
+}
+
+// SyncParentDir fsyncs the parent dir of the given path
+func SyncParentDir(path string) error {
+	return SyncDir(filepath.Dir(path))
+}
+
+// CreateDirIfMissing makes sure that the dir exists and returns whether the dir is empty
+func CreateDirIfMissing(dirPath string) (bool, error) {
+	if err := os.MkdirAll(dirPath, 0755); err != nil {
+		return false, errors.Wrapf(err, "error while creating dir: %s", dirPath)
+	}
+	if err := SyncParentDir(dirPath); err != nil {
```

Comment:
Out of curiosity, do we only need to sync the direct parent of the directory created? What if the `MkdirAll` on line 78 needs to create multiple folders in the path?

Reply:
Yes @caod123 - you are right, this should be on all new folders in the path. I copied this function from existing one where assumption was that up to parent it exists. But better to fix this here. I'll fix this in a separate PR after this.